### ISSUE TITLE
Enhanced AsimovSignificance

### ISSUE THREAD

**Issue by adavidzh** [root-project/root#7998]
Hello everyone,

The [`AsimovSignificance`](https://github.com/root-project/root/blob/87a998d48803bc207288d90038e60ff148827664/roofit/roostats/src/RooStatsUtils.cxx#L61-L81) was a great addition to the arsenal of functions that "just work".

I have recently have noticed that there have been some neat generalisations posted to the [arXiv](https://arxiv.org/abs/2102.04275) and was wondering if it would be a good enhancement.

**Comment by guitargeek** [comment]
Related forum post:
https://root-forum.cern.ch/t/more-asymptotic-formulae/57122/5

**Comment by adavidzh** [comment]
If the project is open to a PR with this, can the team please at least provide some guidelines so that the PR is not wasted time?

**Comment by pcanal** [comment]
What type of guidelines are you asking about? (For PR in general we have: https://root.cern.ch/contribute/ and https://github.com/root-project/root/blob/master/CONTRIBUTING.md )

**Comment by adavidzh** [comment]
> What type of guidelines are you asking about? (For PR in general we have: https://root.cern.ch/contribute/ and https://github.com/root-project/root/blob/master/CONTRIBUTING.md )

_Mais bien sûr._  But for good measure I re-read RTFM to make sure I did not miss any updates:
My question is not about style (commit message phrasing, indentation, capitalization, raw type avoidance, running tests, etc).
I am looking for architectural guidance and for that the Taligent Guide is general very strongly geared towards OOP and ["common sense"](https://root.cern/contribute/coding_conventions/#where-to-go-from-here) is anything but.

To be concrete I am looking for guidance like:
- "use the same namespace and replace AsimovSignificance with a general function that uses internal logic to determine which expression to use" or
- "we should encapsulate this given that the new formulae are a family" or
- "overload AsimovSignificance multiple times taking care that the parameters are named according to these criteria", etc.

Basically @mjbasso ([also here](https://gitlab.cern.ch/mbasso)) [proposed generalisations](https://github.com/mjbasso/asymptotic_formulae_examples) from one signal and one background that we presently have and that means some convention to pass the (arrays or matrices of) parameters and either provide different versions of the functions (as Taligent would probably have it) or process the arguments and choose the formula to use (as is presently done).

It was really a specific question for RooStats maintainers like @guitargeek (also since Basso is in ATLAS) or @lmoneta who have curated the thread and follow the topic in general, not a general question.

**Comment by guitargeek** [comment]
Hello @adavidzh, thank you very much for this initiative!

What you plan to do doesn't get constrained by many of the best practices we have and recommend, because you are planning to add free functions to RooStats. For the addition of classes that potentially get written to disk, we have to evaluate more carefully before what we plan to do, because we have to ensure backwards compatibility in the future.

For adding these utility functions to RooFit, we have only some basic guidelines:

  1. They should be thread safe (i.e. no global state
  2. Design the interface such that it is also natural to use from Python, which often tells us to *not* use overloads, as Python doesn't have this concept. Overloads should only be used if the semantically the arguments are the same, but of different C++ type (e.g. one overload from `std::vector<double>` and one for `TVectorD` or something like that).
  3. For logging, use the [RooMsgService](https://root.cern.ch/doc/master/classRooMsgService.html) facility, no bare `cout` (e.g. `oocoutE(nullptr,InputArguments) << ...` when printing argument errors). 
  4. Prefer C++ standard library over ROOT core classes

Then, we also request a few things to make the development complete:

1. Some unit tests in `roofit/roostats/test`
2. Good doxygen documentation in the `.cxx` code itself
3. Optimally, a tutorial in both Python and C++ that showcases the new functionality to promote adoption
4. No breakage of existing code in ROOT, keeping our `clang-format` style, and address other "requests" from the CI

If you stick to all of that, I'm always happy to merge functionality contributed by RooFits users! Because if it comes from the users themselves, it's relevant to our users 🙂

**Comment by mjbasso** [comment]
Hi @adavidzh @guitargeek - thank you for looking into this! I think it would be very cool if these formulae are implemented in ROOT and see use :-)

I don't have so much useful input on the precise implementation, but some will involve solving systems of equations to yield an estimate of Z0. This is of course still more lightweight than throwing e.g. 100k toys
